src/appservice/Appservice.ts

```diff
@@ -958,16 +958,26 @@ export class Appservice extends EventEmitter {
         }
 
         let responded = false;
-        this.emit("query.key_claim", req.body, async (result: MSC3983KeyClaimResponse | Promise<MSC3983KeyClaimResponse> | undefined | null) => {
-            if (result?.then) result = await result;
-            if (!result) {
-                res.status(404).json({ errcode: "M_UNRECOGNIZED", error: "Endpoint not implemented" });
-                responded = true;
-                return;
-            }
-
-            res.status(200).json(result);
+        this.emit("query.key_claim", req.body, (result: MSC3983KeyClaimResponse | Promise<MSC3983KeyClaimResponse> | undefined | null) => {
             responded = true;
+
+            const handleResult = (result2: MSC3983KeyClaimResponse) => {
+                if (!result2) {
+                    res.status(404).json({ errcode: "M_UNRECOGNIZED", error: "Endpoint not implemented" });
+                    return;
+                }
+
+                res.status(200).json(result2);
+            };
+
+            if ((result as Promise<MSC3983KeyClaimResponse>)?.then) {
+                (result as Promise<MSC3983KeyClaimResponse>).then(r => handleResult(r)).catch(e => {
+                    LogService.error("Appservice", "Error handling key claim API", e);
+                    res.status(500).json({ errcode: "M_UNKNOWN" });
```

Comment:
Could we at least emit an `error` so the consumer has something to go off "Error handling key claim".

Reply:
the consumer ends up swallowing the error as part of the MSC behaviour in both cases, fwiw.